src/dotnet/ToolPackage/PackageId.cs

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Microsoft.DotNet.InternalAbstractions;
+
+namespace Microsoft.DotNet.ToolPackage
+{
+    internal struct PackageId : IEquatable<PackageId>, IComparable<PackageId>
+    {
+        private string _id;
+
+        public PackageId(string id)
+        {
+            _id = id?.ToLowerInvariant() ?? throw new ArgumentNullException(nameof(id));
+        }
+
+        public bool Equals(PackageId other)
+        {
+            return _id == other._id;
+        }
+
+        public int CompareTo(PackageId other)
+        {
+            return _id.CompareTo(other._id);
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null)
+            {
+                return false;
+            }
+
+            return Equals((PackageId)obj);
+        }
+
+        public override int GetHashCode()
+        {
+            return _id.GetHashCode();
```

Comment:
Since this is a struct, default(PackageId).GetHashCode() is possible and would null ref.

Reply:
Will fix.